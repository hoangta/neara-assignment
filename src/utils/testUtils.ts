import { MFA } from "../types/mfa";

export const mockStore = () => {
  jest.mock("../stores/mfaStore", () =>
    jest.fn().mockImplementation(() => ({
      mfas: mockMFAs,
      createMFA: jest.fn(),
      switchMFAs: jest.fn(),
    }))
  );
};

export const mockSystemTime = (date: Date) => {
  jest.useFakeTimers("modern");
  jest.setSystemTime(date);
};

export const mockSVGElementAnimate = () => {
  const circleAnimateSpy = jest.fn();
  SVGElement.prototype.animate = circleAnimateSpy;
  return circleAnimateSpy;
};

export const mockMFAs: MFA[] = [
  {
    provider: "Test Provider",
    code: "012345",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 2),
  },
  {
    provider: "Test Provider 2",
    code: "543210",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 3),
  },
];
