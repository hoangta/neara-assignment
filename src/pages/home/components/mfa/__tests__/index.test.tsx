import React from "react";
import { fireEvent, render } from '@testing-library/react'
import MFAView from '..'
import { mockSVGElementAnimate, mockMFAs } from "../../../../../utils/testUtils";

jest.useFakeTimers("modern");
jest.setSystemTime(new Date(2023, 2, 1));
mockSVGElementAnimate();

describe("Pages/Home/Components/MFAView", () => {
    const mockMFA = mockMFAs[0];
    const onMoveUpSpy = jest.fn();
    const onMoveDownSpy = jest.fn();

    it('should render correctly', () => {
        const { asFragment, getByText } = render(<MFAView mfa={mockMFA} isEditing onMoveUp={onMoveUpSpy} onMoveDown={onMoveDownSpy} />)
        expect(asFragment()).toMatchSnapshot();
        expect(getByText('012 345')).toBeInTheDocument();
    })

    it('should not render MoveBar if isEditing = false', () => {
        const { asFragment, queryByTestId } = render(<MFAView mfa={mockMFA} isEditing={false} onMoveUp={onMoveUpSpy} onMoveDown={onMoveDownSpy} />)
        expect(queryByTestId('moveup_button')).not.toBeInTheDocument();
        expect(queryByTestId('movedown_button')).not.toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot()
    })

    it('should trigger onMoveUp when click on move up button', () => {
        const { getByTestId } = render(<MFAView mfa={mockMFA} isEditing onMoveUp={onMoveUpSpy} onMoveDown={onMoveDownSpy} />)
        const moveUpButton = getByTestId('moveup_button_0');
        fireEvent.click(moveUpButton)
        expect(onMoveUpSpy).toBeCalledTimes(1);
    })

    it('should trigger onMoveDown when click on move down button', () => {
        const { getByTestId } = render(<MFAView mfa={mockMFA} isEditing onMoveUp={onMoveUpSpy} onMoveDown={onMoveDownSpy} />)
        const moveDownButton = getByTestId('movedown_button_0');
        fireEvent.click(moveDownButton)
        expect(onMoveDownSpy).toBeCalledTimes(1);
    })

    it('should render remaining seconds correctly', () => {
        const mockRemainingSeconds = 86400
        const { getByText } = render(<MFAView mfa={mockMFA} isEditing onMoveUp={onMoveUpSpy} onMoveDown={onMoveDownSpy} />)
        expect(getByText(mockRemainingSeconds.toFixed())).toBeInTheDocument();
    })
})