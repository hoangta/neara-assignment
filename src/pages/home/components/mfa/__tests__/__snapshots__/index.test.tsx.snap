// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages/Home/Components/MFAView should not render MoveBar if isEditing = false 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt="Test Provider"
      class="image"
      src="www.testurl.com"
    />
    <div
      class="detail"
    >
      <div
        class="provider"
      >
        Test Provider
      </div>
      <div
        class="code"
      >
        012 345
      </div>
    </div>
    <div
      class="indicatorContainer"
    >
      <svg
        class="indicatorItem"
      >
        <circle
          class="circle"
          cx="20"
          cy="20"
          fill="none"
          r="14"
          stroke="gray"
          stroke-dasharray="87.96459430051421"
          stroke-dashoffset="87.96459430051421"
          stroke-width="2"
        />
      </svg>
      <div
        class="countdownLabel"
      >
        86400
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Pages/Home/Components/MFAView should render correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="moveButtonContainer"
    >
      <button
        class="moveButton"
        data-testid="moveup_button_0"
      >
        ⬆
      </button>
      <button
        class="moveButton"
        data-testid="movedown_button_0"
      >
        ⬇
      </button>
    </div>
    <img
      alt="Test Provider"
      class="image"
      src="www.testurl.com"
    />
    <div
      class="detail"
    >
      <div
        class="provider"
      >
        Test Provider
      </div>
      <div
        class="code"
      >
        012 345
      </div>
    </div>
    <div
      class="indicatorContainer"
    >
      <svg
        class="indicatorItem"
      >
        <circle
          class="circle"
          cx="20"
          cy="20"
          fill="none"
          r="14"
          stroke="gray"
          stroke-dasharray="87.96459430051421"
          stroke-dashoffset="87.96459430051421"
          stroke-width="2"
        />
      </svg>
      <div
        class="countdownLabel"
      >
        86400
      </div>
    </div>
  </div>
</DocumentFragment>
`;
